Add doc for share environments (#193)
scalar-labs/scalar-terraform: docs/ShareEnvironment.md

# How to share environment with others

This explanation is for multiple person operating in the same environment.

## Get a public/private key-pair from the person who built the environment

> :memo: Also need all the tfsatte files when using `local` backend

Reviewer: Small spelling mistake in `tfsatte`.
Expected `tfstate`

Author: Thanks! fixed in e6261d1.

## Update the tfstate of the network module

- Fix key_path of `example.tfvars` in your local

  ```
  public_key_path = "./example_key.pub"

  private_key_path = "./example_key"
  ```

> - Update the output value `private_key_path` and `public_key_path` of tfsate

Reviewer: Small spelling mistake in `tfsatte`.
Expected `tfstate`

Author: ditto.

> - Update the output value `private_key_path` and `public_key_path` of tfstate

Reviewer 2: ```suggestion
Then, update tfstate file as follows.
```

  ```

>   cd example/[aws|azure]/network

Reviewer 2: ```suggestion
  cd /path/to/network-module
```

  terraform init
  terraform refresh -var-file=example.tfvars
  ```

> ## Update the tfstate of the cassandra, scalardl and monitor modules

Reviewer 2: ```suggestion
## Update the tfstate of the cassandra, scalardl, and monitor modules
```

> - Update the value `private_key_path` and `public_key_path` of `data.terraform_remote_state.network` in each tfstate

Reviewer 2: ```suggestion
Update the values of `private_key_path` and `public_key_path` of `data.terraform_remote_state.network` in each tfstate and do as follows.
```

# How to share environments with others

Non-testing environments tend to be shared and operated with multiple DevOps engineers.
This document explains how to do it properly.

> 

Author: @feeblefakie looking good ! 👍

## Get a private key from the person who has built an environment.

You can generate a corresponding public key with `ssh-keygen -yf ~/.ssh/privatekey.pem > publickey.pem`.

If an environment is created with `local` backend, you need all the tfstate files additionally. But, as a best practice, it's better to use cloud storage backends such as AWS S3 or Azure Blob Storage for an environment that is shared among multiple engineers.

## Update the tfstate of the network module

Update the key paths of `tfvars` file in your local such as `example.tfvars` in the example.

```
public_key_path = "./publickey.pem"

private_key_path = "./privatekey.pem"
```

Then, update tfstate file as follows.

```
cd /path/to/network
terraform init
terraform refresh -var-file=example.tfvars
```

## Update the tfstate of the cassandra, scalardl, and monitor modules

> Update the values of `private_key_path` and `public_key_path` of `data.terraform_remote_state.network` in each tfstate and do as follows.

Reviewer 2: I'm testing with psim environment but I don't have tfstate file.
Do we need to pull ?

Author: The tfstate file in the psim environment is on s3, so you don't need to pull it to local.

Reviewer 2: Ok, this line applies only for `local` backend. How about the following commands? They need to be done whether or not local?
Anyways, I think it requires some updates.

Author: >They need to be done whether or not local?

Yes, This is necessary regardless of the back-end type. (not only for local)

> Update the values of `private_key_path` and `public_key_path` of `data.terraform_remote_state.network` in each tfstate and do as follows.

Reviewer 2: ```suggestion
Then, do the following to reflect the above changes to each module.
```

# How to share environments with others

Non-testing environments tend to be shared and operated with multiple DevOps engineers.
This document explains how to do it properly.
## Why we need the following steps

When we build an environment with scalar-terraform, the paths of a specified key pair are stored in tfstate files. Those paths are expanded to the full paths even if they are specified with relative paths, so the paths can be operator dependent variables unless all the operators agree on the same full paths beforehand.
With the following steps, you actually update tfstate files based on your environment to make it work for you.

Note that if you update shared tfstate files in Cloud storage based on your environment, others can not use the tfstate files and they have to do the same steps again to make it be able to work for them. So, it is more like a delegation of the operator in such case.

> ## Why we need the following steps

Reviewer 2: ```suggestion
```

> When we build an environment with scalar-terraform, the paths of a specified key pair are stored in tfstate files. Those paths are expanded to the full paths even if they are specified with relative paths, so the paths can be operator dependent variables unless all the operators agree on the same full paths beforehand.

Reviewer 2: ```suggestion
```

> With the following steps, you actually update tfstate files based on your environment to make it work for you.

Reviewer 2: ```suggestion
```

> Note that if you update shared tfstate files in Cloud storage based on your environment, others can not use the tfstate files and they have to do the same steps again to make it be able to work for them. So, it is more like a delegation of the operator in such case.

Reviewer 2: ```suggestion
```
